<a href="https://colab.research.google.com/github/ridoy1211/Flyrank-Internship-ML/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ridoy1211/Flyrank-Internship-ML/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

Lane: Lane 1 - Ranking Signal Analysis, aimed at a specific, named industry
pattern: the "Great Decoupling" - search impressions holding steady or
rising while click-through rate quietly falls, widely documented since AI
Overviews became standard in Google Search (Ahrefs measured a 58% CTR drop
on top-ranking pages in Feb 2026; Pew Research found only 8% of users click
through when an AI summary appears, vs. 15% without one).

I'm choosing this because it's a real, currently escalating business fear,
not a hypothetical one: a page can look completely healthy on impressions
and rank while its actual value is quietly disappearing, and most reporting
dashboards don't surface that on their own. The published research on this
is all aggregate/industry-level. Nobody has shown, from clean page-level
data, which content characteristics make a page MORE or LESS exposed to
this pattern - which is the specific, narrower, and more original question
I want this capstone (and eventual paper) to answer.

In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
pass

## 2. The question: decision, action, cost of a wrong call

Question: Which observable page characteristics (intent, content type,
position, freshness) are associated with a "Great Decoupling" signature —
impressions holding steady while clicks meaningfully decline over a recent
window — and can this be turned into an early-warning exposure signal?

Unit of analysis: one content page (content_id), evaluated using its
before/after 30-day windows.

Decision this improves: which pages an SEO/content strategist flags for
defensive review in a quarterly portfolio triage, before the click loss is
large enough to show up as an obvious revenue problem in a normal report.

Who acts, and how: a content or SEO strategist, using the flagged pages and
their reason codes to prioritize defensive work (e.g. adding original data,
restructuring for AI-citation capture, strengthening why a user should
still click) instead of routine "this page is old" refreshes.

Cost of a wrong call: false positive — the team spends scarce strategist
time defending a page whose click drop was really seasonality or a sibling
page absorbing demand, not decoupling. False negative — a real,
still-valuable page keeps silently losing clicks unnoticed, and the
eventual revenue loss is larger and harder to reverse than if it had been
caught early.

Why data/ML helps: this pattern is invisible to a normal glance at
impressions or rank alone — it only appears by deliberately comparing two
time windows against each other across thousands of pages, which is
exactly the kind of systematic check that's impractical by hand but
straightforward to compute and rank.

In [4]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
pass

In [5]:
import os

# If the data file isn't here, we're likely in Colab without the repo cloned — clone it.
if not os.path.exists('data/raw/content_refresh_anonymized.csv'):
    if not os.path.exists('Flyrank-Internship-ML'):
        !git clone https://github.com/ridoy1211/Flyrank-Internship-ML.git
    os.chdir('Flyrank-Internship-ML')

print("Working directory:", os.getcwd())
print("File exists:", os.path.exists('data/raw/content_refresh_anonymized.csv'))

Working directory: /content/Flyrank-Internship-ML
File exists: True


## 3. Quick look at the data (2-3 real numbers)
Loading data/raw/content_refresh_anonymized.csv to check whether AI-referral
patterns exist at all in this data, and how sparse they really are.

In [6]:
import pandas as pd

df = pd.read_csv('data/raw/content_refresh_anonymized.csv')
n = len(df)

# only pages with real signal in both windows (avoids divide-by-zero noise)
sub = df[(df['impressions_prev_30d'] > 0) & (df['clicks_prev_30d'] > 0)].copy()

sub['impr_change_pct'] = 100 * (sub['impressions_last_30d'] - sub['impressions_prev_30d']) / sub['impressions_prev_30d']
sub['click_change_pct'] = 100 * (sub['clicks_last_30d'] - sub['clicks_prev_30d']) / sub['clicks_prev_30d']

# the signature: impressions roughly flat, clicks meaningfully down
pattern = sub[(sub['impr_change_pct'].between(-10, 10)) & (sub['click_change_pct'] <= -15)]

print(f"usable rows (real prior-window data): {len(sub)} ({100*len(sub)/n:.1f}% of {n})")
print(f"Great Decoupling-signature rows: {len(pattern)} ({100*len(pattern)/len(sub):.1f}% of usable rows)")
print(f"clients affected: {pattern['client_id'].nunique()} of {sub['client_id'].nunique()}")
print(f"median impressions_90d, affected vs all usable: {pattern['impressions_90d'].median()} vs {sub['impressions_90d'].median()}")

usable rows (real prior-window data): 11759 (39.2% of 30000)
Great Decoupling-signature rows: 674 (5.7% of usable rows)
clients affected: 22 of 29
median impressions_90d, affected vs all usable: 4232.5 vs 4686.0


Result: 674 of 11,759 usable pages (5.7%) show the signature, spread
across 22 of 29 clients, with a median impressions_90d (4,232) nearly
identical to the portfolio-wide median (4,686) — meaning these are pages
with real, meaningful traffic, not dying long-tail pages. That's enough
evidence this pattern is real and worth building on.

## 4. Careful words: what I can and can't claim

What this can say: which pages show the observed "impressions flat, clicks
down" signature over a real before/after window, how common that signature
is in this portfolio, and (later, once tested) which content
characteristics predict it - an observed, decision-support pattern, backed
by published industry research on the broader Great Decoupling trend.

What this can never say: that any single page's click drop was CAUSED by an
AI Overview specifically - this data can't distinguish that from a changed
SERP layout, seasonality, or a sibling page absorbing demand
(consolidation). Ruling those look-alikes out is my next step, not an
assumption I get to skip. I also won't claim to have measured Google's or
any AI system's algorithm directly — only observable page-level signals,
and I'll cite the external, aggregate industry research (Ahrefs, Pew) as
context, not as proof about my own dataset.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
pass

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.